libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,182 @@
+
+#include <libevmasm/AssemblyItem.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/slice.hpp>
+
+#include <optional>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::evmasm;
+
+bool Inliner::isInlineCandidate(u256 const& _tag, InlinableBlock const& _block) const
+{
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+
+	// Never inline tags that reference themselves.
+	for (AssemblyItem const& item: _block.items)
+		if (item.type() == PushTag)
+			if (_tag == item.data())
+				return false;
+
+	return true;
+}
+
+map<u256, Inliner::InlinableBlock> Inliner::determineInlinableBlocks(AssemblyItems const& _items) const
+{
+	std::map<u256, ranges::span<AssemblyItem const>> inlinableBlockItems;
+	std::map<u256, uint64_t> numPushTags;
+	std::optional<size_t> lastTag;
+	for (auto&& [index, item]: _items | ranges::views::enumerate)
+	{
+		// The number of PushTag's approximates the number of calls to a block.
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, true))
+		{
+			if (item == Instruction::JUMP)
+				inlinableBlockItems.emplace(
+					piecewise_construct,
+					forward_as_tuple(_items[*lastTag].data()),
+					forward_as_tuple(_items | ranges::views::slice(*lastTag + 1, index + 1))
+				);
+			lastTag.reset();
+		}
+
+		if (item.type() == Tag)
+			lastTag = index;
+	}
+
+	// Filter candidates for general inlinability and store the number of PushTag's alongside the assembly items.
+	map<u256, InlinableBlock> result;
+	for (auto&& [tag, items]: inlinableBlockItems)
+		if (uint64_t const* numPushes = util::valueOrNullptr(numPushTags, tag))
+		{
+			InlinableBlock block{items, *numPushes};
+			if (isInlineCandidate(tag, block))
+				result.emplace(std::make_pair(tag, InlinableBlock{items, *numPushes}));
+		}
+	return result;
+}
+
+namespace
+{
+optional<AssemblyItem::JumpType> determineJumpType(AssemblyItem::JumpType _intoBlock, AssemblyItem::JumpType _outOfBlock)
+{
+	auto jumpTypeToInt = [](AssemblyItem::JumpType _jumpType) -> int {
+		switch (_jumpType)
+		{
+			case AssemblyItem::JumpType::IntoFunction:
+				return +1;
+			case AssemblyItem::JumpType::OutOfFunction:
+				return -1;
+			case AssemblyItem::JumpType::Ordinary:
+				return 0;
+		}
+		assertThrow(false, OptimizerException, "");
+	};
+	switch (jumpTypeToInt(_intoBlock) + jumpTypeToInt(_outOfBlock))
+	{
+		case 0:
+			return AssemblyItem::JumpType::Ordinary;
+		case 1:
+			return AssemblyItem::JumpType::IntoFunction;
+		case -1:
+			return AssemblyItem::JumpType::OutOfFunction;
+		default:
+			return nullopt;
+	}
+}
+}
+
+optional<AssemblyItem> Inliner::shouldInline(u256 const&, AssemblyItem const& _jump, InlinableBlock const& _block) const
```

Comment:
In this function we should have all information available now, that we may need for any kind of complex heuristics I think.

We have the jump to the tag that we want to inline (and thereby its jump type), we know all the opcodes to be inlined (including the exit jump and its jump type), and we know how often the tag is pushed throughout the bytecode.

So we could actually now really calculate the cost of inlining vs not inlining with a proper gas estimate using the ``runs`` parameter...

So we have to decide if we want to do this right way or keep it simple for now - keeping in mind that this might affect which standard-json options we want to define for it.

Reply:
One thing we could actually additionally determine by walking the items one more time up front: how many of the ``PushTag``s we've seen actually have the form ``PUSHTAG JUMP``, i.e. can *actually* be inlined.

Reply:
Inlining can result in further optimizations, so computing the actual gas tradeoff is not perfect, but could certainly help.